llvm/lib/Target/AMDGPU/SIInstrInfo.cpp

```diff
@@ -5094,6 +5094,39 @@ unsigned SIInstrInfo::getVALUOp(const MachineInstr &MI) const {
       "Unexpected scalar opcode without corresponding vector one!");
 }
 
+bool SIInstrInfo::isSCCDefinedBefore(MachineBasicBlock &MBB,
```

Comment:
These functions are just recreating computeRegisterLiveness. However, I don't think you really need that in the first place. You could just unconditionally copy and restore and let PeepholeOpt deal with it later if it was unnecessary

Reply:
When I did not check for liveness, there were save/restore sticking around. Maybe we need to fix peephole to fix those.

Reply:
I guess PeepholeOpt can't deal with the intermediate def. You still should use computeRegisterLiveness

Reply:
I looked at computeRegisterLiveness code, and I am not sure if that satisfies what I want to do. 
An inst may set SCC to be live but the result of which may never be used. In that case, I don't want to save/restore. 
computeRegisterLivess always goes backwards first and checks for "defs", and if it finds defs, it says it's live. That is correct. That is necessary but not sufficient reason to generate save/restore of SCC.

Reply:
We could change computeRegisterLiveness to change to do the following:
1. Default as is right now.
2. Just do backward scan to defs
3. Just do forward scan to uses.

Reply:
The reason for using computeRegisterLiveness is that handles edges case such as liveins/liveouts, bundles, etc.

We should just trust the dead flag here.  The case you are trying to capture should presumably already be handled by this?  A def before the waterfall without a use after it, should be dead. In any other scenario we should consider SCC to be live.

Preserving SCC if `computeRegisterLiveness(Reg) != LQR_DEAD` should be sufficient.  As computeRegisterLiveness will check the dead flags and check for clobbering during its forward search.

Reply:
Pushing a new patch using `computeRegisterLiveness(Reg)`